File: backend/internal/domain/interfaces.go
package domain

import (
	"context"

	"github.com/kenyamamoto/pokelingual/backend/internal/model"
)

// AIScorer scores English-to-Japanese translations and provides chat capabilities.
// Implement this interface to swap between Gemini, Claude, OpenAI, etc.
type AIScorer interface {
	ScoreTranslation(ctx context.Context, englishText, japaneseTranslation string) (*model.ScoreResult, error)
	Chat(ctx context.Context, chatCtx *model.ChatContext, messages []model.ChatMessage) (string, error)
}

// PokemonFetcher fetches Pokemon data from an external source.
type PokemonFetcher interface {
	GetRandomPokemon(ctx context.Context) (*model.Pokemon, error)
	GetPokemonByID(ctx context.Context, id int) (*model.Pokemon, error)
}

// UserPokemonRepository manages user's Pokemon collection data.
type UserPokemonRepository interface {
	UpsertEncounter(ctx context.Context, uid string, pokemonID int, score float64, captured bool) error
	GetCollection(ctx context.Context, uid string) ([]model.UserPokemon, error)
	GetPokemon(ctx context.Context, uid string, pokemonID int) (*model.UserPokemon, error)
}

// UserSettingsRepository manages per-user preferences such as excluded Pokemon.
type UserSettingsRepository interface {
	GetSettings(ctx context.Context, uid string) (*model.UserSettings, error)
	UpdateExcludedPokemon(ctx context.Context, uid string, pokemonIDs []int) error
}

File: backend/src/domain/interfaces.ts
import type {
  ScoreResult,
  ChatContext,
  ChatMessage,
  Pokemon,
  UserPokemon,
  UserSettings,
} from "../types/index.js";

/** 翻訳採点と教授チャットを担う AI スコアラ。GeminiService と MockAIScorer が実装する。 */
export interface AIScorer {
  scoreTranslation(englishText: string, japaneseTranslation: string): Promise<ScoreResult>;
  replyToChat(chatCtx: ChatContext, messages: ChatMessage[]): Promise<string>;
}

/** ポケモン情報の取得を担うフェッチャ。PokeAPIService と MockPokemonFetcher が実装する。 */
export interface PokemonFetcher {
  getRandomPokemon(): Promise<Pokemon>;
  getPokemonByID(id: number): Promise<Pokemon>;
}

/** ユーザの図鑑進捗 (遭遇/捕獲) を永続化するリポジトリ。 */
export interface UserPokemonRepository {
  upsertEncounter(uid: string, pokemonID: number, score: number, captured: boolean): Promise<void>;
  getCollection(uid: string): Promise<UserPokemon[]>;
  getPokemon(uid: string, pokemonID: number): Promise<UserPokemon>;
}

/** ユーザ設定 (除外ポケモン等) を永続化するリポジトリ。 */
export interface UserSettingsRepository {
  getSettings(uid: string): Promise<UserSettings>;
  updateExcludedPokemon(uid: string, pokemonIDs: number[]): Promise<void>;
}

/** 当日のレートリミット利用状況。count が消費数、limit が上限値。 */
export interface DailyUsage {
  count: number;
  limit: number;
}

/** 日次レート制限カウンタを管理するリポジトリ。Firestore 版とインメモリ版がある。 */
export interface RateLimitRepository {
  checkAndIncrement(uid: string): Promise<DailyUsage>;
  getUserUsage(uid: string): Promise<DailyUsage>;
}
